fsm.go

```diff
@@ -135,6 +144,18 @@ func (fsm *fsm) State() (consensus.State, error) {
 	return dupState(fsm.state)
 }
 
+func (fsm *fsm) updateSubscribers(st consensus.State) {
+	fsm.chMux.Lock()
+	defer fsm.chMux.Unlock()
+	if fsm.subscriberCh != nil {
+		select {
+		case fsm.subscriberCh <- st:
+		default:
+			logger.Error("subscriber channel is full. Discarding state!")
```

Comment:
Every individual state is the resultant state at that point (I think its your former case). The state is not thrown away in any case, it can always be fetched with `GetCurrentState()`. This is just a convenient way to inform that a state update happened.

I would leave it to the consumer to ensure that it can consume that channel and just drop updates otherwise (like it happens now). If we start caching too much because a bad consumer doesn't read anything we need to do more sophisticated memory management etc. The consumer can always find out the current state anyway. This may be a totally different discussion for OpLogConsensus though, where every update matters as you point out.

I hope I have explained it correctly... Does that make any sense?